### DISCUSSION TITLE
Welcome to nexus433 Discussions!

### DISCUSSION THREAD

**Discussion by aquaticus** [aquaticus/nexus433#25] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.

**Comment by SmokeDayz** [comment]
I am trying to use your solution to read some temperature sensors but not having much luck - I am a noob! I launch the program and it does not see any sensors. Not sure what to try next - here is some background:
- using raspberry pi 4 with superheterodyne receiver
- tested receiver with rpi-rf and there is tons of traffic showing up
- I do not have home assistant setup - I wanted to see if I could get this working first
- I wasn't sure how I would go about downloading the appropriate package so i built it using your manual build installation
- mosquito has been installed and is running on the same pi as nexus 433
- i can see this:
Loading configuration from /etc/nexus433.ini
Reading data from the 433MHz receiver on /dev/gpiochip0 pin 17.
Decoder resolution: 1 µs; tolerance: 300 µs
Connected to MQTT broker.

Is it possible that the sensors are not compatible? I have weather sensors that were decodable in SDR
Any help would be great! Thanks for sharing your work!

**Reply by aquaticus** [reply]
First, I would double check obious things:
* if the received is powered
* if it's connected to pin 17
 
If you sure, the receiver is connected properly most likely your sensors are using different protocol.
SDR can decode a lot of different protocols while this software only one.

**Reply by JTkode** [reply]
Hi, 
Thanks for creating this awesome application! Seemed to fit my project like a glove!
But ran into a similar problem as Smoke above. Same line of text (MQTT broker...) and it just hangs there.
Receiver is connected to GPIO17 and is getting power. Tried 2 different antennas.
Transmitter is however different than yours (https://github.com/marcelm/clas-ohlson-36-6726) but I thought the protocol should be similar.. same bits, same intervals etc... 
Can you help? is the protocol different?

**Reply by aquaticus** [reply]
It may be a little bit different and cannot be decoded. SDR supports much more dfferent sensors.
First I would check if you got ny signal on GPIO17 using oscilloscope or logic analyser. If you see something that looks like transmission that means protocol is different. If you see nothing probably something wrong with the receiver.

As an alternative you can use https://github.com/merbanan/rtl_433 it works pretty well with SDR/DVB cheap receivers.

**Reply by JTkode** [reply]
Thank you for a really quick reply! 
While I'm waiting for that DVB dongle to arrive, I tried to modify your code and get it to work that way. But I was wondering.
When this was installed to my rasberry it created 2 folders.. nexus433 and release. I modified the "decoder.cpp" in the nexus433 folder to match the microseconds for the transmitter, then ran the 'make -j' command and install as you stated was needed. Then 'sudo nexus433 --verbose -n 17' but nothing, a same thing as above.
Question is, were my modifications included in the "make -j" -command in that nexus433 folder or do I have to edit them elsewhere and then run the makes and installs?
Mods i did in decoder.cpp: 
500----->no change
1000--->850
2000--->1800
4000--->3800

**Reply by aquaticus** [reply]
When you compile the code (`make -j`) the result is stored in `release` directory. You can run the program directly from this location (e.g. `sudo ./release/nexus433 --verbose -n 17`). This is faster and usual way of testing the changes. You can install the program later when you're sure it works as expected.

Make sure you run your new version of the program not the old one (you can add say `printf("Hello\n")` in `main()` function.

Note that your modification maybe not enough to make it working, so you won't see any difference (still not working). I saw that some sensors send 35 bits instead of 36. The best way is to verify receiver output on the scope and measure timings.

**Reply by JTkode** [reply]
Thank you for your help and really quick replies! 
Some updates to go on. I tried with different receiver and I did get a verification that the transmitters do actually send the exact same bits you have discribed. After that I got the original receiver to work and your code did read these clas ohlson sensors so I think you can add these here as well as supported :)  (nexus433 --verbose -n 17) command works -> shows all sensors!! 
But I could not and cannot somehow get this nexus433 command to read my modifications. I do not know where does it run from. I've tried to edit the main.cpp from nexus433 folder but nothing I do there has any effect on anything. I've tried that printf("hello n\") in the main.cpp but also I have edited the "help" string text with my own text ( you can quess what words they were) and then running the make -j in the release folder and nothing. I sees the modified files and compiles them but running nexus433 command ignores them..
My goal is to get simple output file with updated lines from the sensors.. but I cannot test anything as my edits have no effect :/

**Reply by aquaticus** [reply]
`nexus433 --verbose -n 17` executes the program that is located in `/usr/local/bin`.
To run your compiled version you must specify exact path to compiler output, if you're in `debug` directory call `./nexus433 --verbose -n 17` (`./` cannot be skipped).

**Reply by JTkode** [reply]
Thank you for your help, once again.
I know this is far from the "original problem" which was caused by [not grounding the fourth on/off pin in the 433 module] but I hope if it's ok to continue under this discussion, if not I can open another question feed.
Anyway, I've reached my previous goal to get the data to a file but now i'm stuck with getting nexus to run in the background. I tried to make it as a service but it didn't update the file and behaved differently than before. My code ran fine under the verbose -command but didn't run when service was called. 
I thought that the service would read and call the it->Get(what eva) variable in the main.cpp as normal and then would output them for the MQTT to use but this is not the case? So how does the sudo service actually work? Where is that run from? What variables it uses? 
I've tried to add print this and that to the mix but ofcourse something broke. Sudo service doesn't start and it somehow locks the nexus433.ini file so that the normal nexus433 --verbose don't work (says it doesn't find it etc...) unless I reboot the pi. Is there a way to unluck this ini file without restart?

**Reply by aquaticus** [reply]
No sure I understand you correctly.
* `sudo` is a command to run program with root privileges. For nexus433 you need it to read pin state.
* you can unlock INI file by killing the process that keeps the file. First call `ps -A`, note process id of nexus433 and execute `kill -9 proc_id`
* nexuss433 should work in the same way as from command line. Not sure what `sudo service` means, you start service by `systemctl start nexus433` . Please not that there was a bug introduced in one of the latest commits that prevented nexus433 to start in service mode.

**Reply by JTkode** [reply]
Thank you for your help and patience ..again :)
I'm still stuck with this service part. I do not understand while it's running (atleast ithink it's running) how it bypasses nearly all the code in the main.cpp. Only when status or stop is initiated, something happens from the code. 

This seems to show that it is running...
pi@raspberrypi:~/release $ sudo service nexus433 status
● nexus433.service - 433MHz sensor receiver
   Loaded: loaded (/etc/systemd/system/nexus433.service; enabled; vendor preset: enabled)
   Active: active (running) since Sat 2022-10-29 13:59:55 EEST; 6s ago
 Main PID: 32506 (nexus433)
   CGroup: /system.slice/nexus433.service
           └─32506 /usr/local/bin/nexus433

Oct 29 13:59:55 raspberrypi systemd[1]: Started 433MHz sensor receiver.
Oct 29 13:59:55 raspberrypi nexus433[32506]: Loading configuration from /etc/nexus433.ini
Oct 29 13:59:55 raspberrypi nexus433[32506]: Reading dataVITTU2 from 433MHz receiver on /dev/gpiochip0 pin 27. K3 0 tmp
(->there is some frustration visible here and the zero is temperature variable from verbose part of the code.. shows nothing here but when service is stopped and status is read, it gets the temperature and shows it. )


When service statuses are read, nexus433 doesn't show up.

pi@raspberrypi:~/release $ sudo service --status-all
 [ - ]  alsa-utils
 [ - ]  apparmor
 [ + ]  avahi-daemon
 [ - ]  bluetooth
 [ - ]  console-setup.sh
 [ + ]  cron
 [ + ]  dbus
 [ + ]  dhcpcd
 [ + ]  dphys-swapfile
 [ + ]  fake-hwclock
 [ - ]  fio
 [ - ]  hwclock.sh
 [ - ]  keyboard-setup.sh
 [ + ]  kmod
 [ + ]  lightdm
 [ + ]  mosquitto
 [ + ]  networking
 [ - ]  nfs-common
 [ - ]  paxctld
 [ - ]  plymouth
 [ - ]  plymouth-log
 [ + ]  procps
 [ + ]  raspi-config
 [ ? ]  rng-tools
 [ - ]  rpcbind
 [ - ]  rsync
 [ + ]  rsyslog
 [ - ]  screen-cleanup
 [ + ]  ssh
 [ - ]  sudo
 [ + ]  triggerhappy
 [ + ]  udev
 [ - ]  x11-common

I checked the bug thing and in my nexus433.service file it was correct.
So should the "service" -mode bypass the main.cpp codes and is the code ran from somewhere else? 
I hope you can help.. i've run out of ideas to try..

**Comment by vadelma-pi** [comment]
Dear aquaticus, thanks a lot for the Nexus433 project. I have used it before and found extremely useful for communicating between Rpi3 and 433 temperature sensor.

Now I want to re-install it on another Rpi3, but run into errors:
1. If I install it form nexus433_1.1.0_raspberrypi_armhf.deb: The following packages have unmet dependencies:
 nexus433 : Depends: libgpiod2 but it is not installable

2. Using manual build:
-- Software version: v1.1.0-1-gf089285
chmod: changing permissions of '/home/pi/release/build_debian_package.sh': Operation not permitted
[ 12%] Built target version
[ 12%] Building CXX object CMakeFiles/nexus433.dir/config.cpp.o
In file included from /home/pi/nexus433/xini.h:24:0,
                 from /home/pi/nexus433/config.cpp:21:
/home/pi/nexus433/INIReader.h: In function ‘char* strncpy0(char*, const char*, size_t)’:
/home/pi/nexus433/INIReader.h:164:32: error: unknown option after ‘#pragma GCC diagnostic’ kind [-Werror=pragmas]
 #pragma GCC diagnostic ignored "-Wstringop-truncation"
                                ^~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
CMakeFiles/nexus433.dir/build.make:62: recipe for target 'CMakeFiles/nexus433.dir/config.cpp.o' failed
make[2]: *** [CMakeFiles/nexus433.dir/config.cpp.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/nexus433.dir/all' failed
make[1]: *** [CMakeFiles/nexus433.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2

Would you have a hint how to overcome these errors?
Thank you.

**Reply by aquaticus** [reply]
Hi,
first of all try to install manually `libgpiod2`.
```bash
sudo apt update
sudo apt install -y libgpiod-dev
```
After that try to install nexus433 package:
```bash
sudo apt install ./nexus433_1.0.2-raspberrypi_armhf.deb
```

For manual install just comment out line 164 in `INIReader.h`. 
```C
//#pragma GCC diagnostic ignored "-Wstringop-truncation"
```
gcc version you are using does not know this pragma (by the way what version of gcc you got on RPI?).
Try to recompile.

**Reply by vadelma-pi** [reply]
Dear aquaticus, 
Thanks a lot for a quick reply!

* The first advice regarding installation through the debian package didn't work. I already had the `libgpiod-dev` of the latest version and the error remains - probably because in the name of the package included in libgpiod is `libgpiod1`?
* But the second advice to comment out line 164 in `INIReader.h` worked well, and now I have Nexus433 is installed using `make` and works fine. Thanks again!
* I have RPI3 with `gcc (Raspbian 6.3.0-18+rpi1+deb9u1) 6.3.0 20170516`

**Reply by aquaticus** [reply]
good to hear you fixed the problem.

Issue with insatllation is relaeted to `libgpiod1`. Debian package depends on `libgpiod2. You could fix that by modifying on one line https://github.com/aquaticus/nexus433/blob/4d72d32b1c39cdcce43df525ae5f2cc3ef6bdd0b/debian_package.cmake#L3
